Threads: Id and mutations

elbotho opened the pull request:
- reworks existing mutation to follow new guidelines
- adds mutations for `setThreadArchived`, `setThreadState` and `setCommentState` and tests
- adds placeholder for @CarolinJaser 's `createComment` code
- remove `trashed` and `updatedAt` from threads
- adds an `id` (base64 of t+firstCommentId) to threads

TODO:
- [x] Add contract test (related: #46)
- [x] improve tests
- [x] implement createComment
- [x] add cache logic to new mutations
- [x] move data into payloads for mutations

backlog:
- (for later PR) Filter for trashed threads/comments

elbotho commented:
Hej @kulla and @inyono could one of you review this and then do a short session with me for 2-3 remaining questions/problems?
It's a pretty big one and I'm sure I a lot of things to improve.

kulla commented on src/internals/graphql/resolver.ts:
```diff
@@ -39,6 +39,8 @@ export type MutationResolver<A, T> = Resolver<
     ? Omit<T, 'query'> & { query: Record<string, never> }
     : T
 >
+export type OverwriteRecord<T, R> = Omit<T, 'record'> & { record: R | null }
```
```suggestion
export interface MutationResponse {
  success: boolean,
  query: Record<string, never>
}

export interface MutationResponseWithRecord<R> extends MutationResponse {
  record: R
}
```

This might be easier to use for others.

kulla commented on src/schema/uuid/thread/types.ts:
```diff
@@ -43,11 +58,10 @@ export interface ThreadData {
 
 export interface ThreadResolvers {
   Thread: {
+    id: Resolver<ThreadData, never, string | null>
```
```suggestion
    id: Resolver<ThreadData, never, string>
```

Property is `id: String!`. Thus it needs to return a string.

kulla commented on src/schema/uuid/thread/utils.ts:
```diff
@@ -74,3 +74,11 @@ export function createThreadResolvers(): ThreadAwareResolvers {
+}
+
+export function decodeThreadId(threadId: string) {
+  return parseInt(Buffer.from(threadId, 'base64').toString('ascii').substr(1))
```
```suggestion
  return parseInt(Buffer.from(threadId, 'base64').toString('utf8').substr(1))
```

elbotho replied:
phew, good catch.

kulla commented on src/schema/uuid/thread/utils.ts:
```diff
@@ -74,3 +74,11 @@ export function createThreadResolvers(): ThreadAwareResolvers {
+  return Buffer.from(`t${firstCommentId}`).toString('base64')
+}
+
+export function decodeThreadId(threadId: string) {
+  return parseInt(Buffer.from(threadId, 'base64').toString('ascii').substr(1))
+}
```
```suggestion
export function decodeThreadId(threadId: string): number | null {
  const result = parseInt(Buffer.from(threadId, 'base64').toString('ascii').substr(1))
  return Number.isNaN(result) || result <= 0 ? null : result
}
```

With this one cannot forget to check, whether a passed id is a valid id or not since the typescript compiler will complain when `number | null` is used in a case where `number` is necessary. Example:

```ts
const threadId = "..."
const id = decodeThreadId(threadId)

// will complain since id = null is not allowed, 
// so you don't forget to handle the case where `threadId` is not valid
await dataSources.serlo.setThreadArchive({ id }) 
```

kulla replied:
The option monad might be useful here: https://dev.to/waynevanson/fp-ts-meets-if-else-and-switch-part-1-3-2lpf

elbotho replied:
Did it with the suggested `number | null` for now. Let's talk about it.

kulla commented on src/schema/uuid/thread/resolvers.ts:
```diff
@@ -20,31 +20,32 @@
   Thread: {
-    createdAt(thread, _args) {
-      return thread.commentPayloads.map((comment) => comment.date).reduce(R.min)
+    id(thread, _args) {
```
```suggestion
    id(thread) {
```

kulla commented on src/schema/uuid/thread/resolvers.ts:
```diff
@@ -20,31 +20,32 @@
+    createdAt(thread, _args) {
+      return thread.commentPayloads[0].date
     },
     title(thread, _args) {
```
```suggestion
    title(thread) {
```

in other places as well...

kulla commented on src/schema/uuid/thread/types.graphql:
```diff
@@ -1,9 +1,8 @@
 type Thread {
+  id: String!
   createdAt: DateTime!
-  updatedAt: DateTime!
```
With the new comment system it will be possible to edit a comment inside a thread. To display this in the new frontend: Where do you need the `updatedAt` property?

elbotho replied:
Could not think of a good use for that on Threads so I removed it. 
Do you agree?

kulla commented on src/schema/uuid/thread/types.graphql:
```diff
@@ -66,5 +65,68 @@ type CommentEdge {
+}
+
+input ThreadCreateCommentInput {
+  title: String!
```
```suggestion
```

An answer does not have a title

elbotho replied:
Caros turf.

kulla commented on src/schema/uuid/thread/types.graphql:
```diff
@@ -66,5 +65,68 @@ type CommentEdge {
+}
+
+input ThreadSetThreadArchivedInput {
+  id: String!
```
```suggestion
  threadId: String!
```

for consistency with `createComment()` (or we use `id` in both endpoints)

elbotho replied:
Hm, I chose `id` here to be consistent with all other `set…` mutations and `threadId` above to make sure the it's clear that you need to input the Id of a thread here. Let's discuss this.

kulla commented on src/schema/uuid/thread/types.graphql:
```diff
@@ -66,5 +65,68 @@ type CommentEdge {
+  setThreadState(
+    input: ThreadSetThreadStateInput!
+  ): ThreadSetThreadStateResponse
+  setCommentState(
```
@elbotho For the frontend it is better to not use `uuid.setState()` here, right?

elbotho replied:
Why I did it this way:
- I remember Jonas saying he'd prefer a distinct endpoint
- from the consumer (frontend) perspective I'd say it's clearer this way (don't need to know if threads/comments are uuids or not)
- this way we can easier change the implementation without changing the endpoints

kulla commented on __tests__/schema/uuid/thread.ts:
```diff
@@ -567,9 +776,9 @@ function setupThreads(uuidPayload: UuidPayload, threads: CommentPayload[][]) {
 }
 
-function createAddCommentMutation(id: number, date: string) {
+function setupComments(id: number, date: string) {
```
Move this into the describe block of `createThread` at the end (so it's closer to where the function is used).

kulla reviewed (changes requested):
First feedback (I'll add new feedback later)

- [ ] a test for the new property `Thread.id` (it is sufficient to check that the result is a string)

elbotho commented on src/schema/uuid/thread/resolvers.ts:
```diff
@@ -80,15 +81,80 @@ export const resolvers: ThreadResolvers = {
+        ...payload.input,
+        userId,
+      })
+      console.log(commentPayload)
```
placeholder code, log stops TS from complaining ;)

kulla commented on src/schema/uuid/thread/resolvers.ts:
```diff
@@ -80,15 +81,85 @@ export const resolvers: ThreadResolvers = {
+          query: {},
+        }
+      const res = await dataSources.model.serlo.setUuidState({
+        id: [firstCommentId],
```
```suggestion
        ids: [firstCommentId],
```

Since it is a list, the parameter name should be changed

kulla commented on src/schema/uuid/thread/resolvers.ts:
```diff
@@ -80,15 +81,85 @@ export const resolvers: ThreadResolvers = {
+    async setCommentState(_parent, payload, { dataSources, userId }) {
+      assertUserIsAuthenticated(userId)
+      const { id, trashed } = payload.input
+      const res = await dataSources.model.serlo.setUuidState({
+        id: [id],
+        userId,
+        trashed,
+      })
+      return {
+        success: res[0] !== null,
+        query: {},
+      }
```
That's the same code as above. You can move this into a helper function.

elbotho replied:
I'd like to keep it for now, since this code might be different again soon.

kulla commented on src/schema/uuid/thread/types.ts:
```diff
@@ -60,7 +73,29 @@ export interface ThreadResolvers {
+    >
+    createComment: MutationResolver<
+      ThreadMutationCreateCommentArgs,
+      ThreadCreateCommentResponse
```
```suggestion
      MutationResponse
```

Then you do not need the `... ? ... : ...` magic in the type `MutationResolver`

elbotho replied:
Good idea.
Okay if we keep it until we are a bit more that the Response types won't change so soon?

kulla replied:
fine for me.

kulla commented on src/model/serlo.ts:
```diff
@@ -465,6 +471,20 @@ export function createSerloModel({
+      const id = decodeThreadId(payload.id)
+      if (id === null) return null
+      return await post<CommentPayload | null>({
+        path: `/api/archive-comment/${id}`,
```
We need to discuss whether we want to pass the argument `id` in the url or in the body payload. For mutation I would pass the id in the payload because then it is easier to change it into in array in the future and we pass all arguments in the same way.

elbotho replied:
warning: changed `setUuidState` and `setNotificationState` as well.
@kulla could you make sure to tell me to revert this here or prepare the legacy endpoints for that?

inyono replied:
> We need to discuss whether we want to pass the argument

IMHO: body all the way (I find it super strange for POST requests to do some stuff via body and some via URL).

elbotho replied:
@inyono nice, that's what we discussed and what I implemented. Match!

kulla reviewed (changes requested):
Here is my feedback so far.

elbotho commented on __tests__/schema/uuid/thread.ts:
```diff
@@ -281,22 +258,22 @@ describe('uuid["threads"]', () => {
       variables: { id: article.id },
       data: {
-        uuid: { threads: { nodes: [{ archived: comment1.archived }] } },
+        uuid: { threads: { nodes: [{ id: encodeThreadId(comment1.id) }] } },
```
does not seem to work when using `assertSuccessfulGraphQLQuery`…

elbotho commented:
@kulla Okaaay:

**Changes since our last session:**
- added rudimentary cache stuff: should invalidate relevant cache keys
- added contract tests for the new mutations (might solve #46, please check)
- integrated your suggestions
- moved mutation tests into own files
- removed all URL data for all mutations (moved ids to body)

**What did not happen :)**
- Tests do not test cache at all (I figure we want to add correct cache mutations soon anyway)
- Two tests should return with 400 errors but return `INTERNAL_SERVER_ERROR` (both are called `mutation is unsuccessful for non existing id`)
- Caros comment code is not commited here
- Did not write super amazing test helpers


That's probably as far as I get atm. Do you think this is a good point for you to take over?
I guess checking cache logic and if the tests are ok would be enough for this to finally be merged.
(Finishing comment mutation could maybe wait?)

elbotho commented:
Changes some things to make `createComment` work.
Added contract test.

Missing: Cache logic in `serlo.ts` (see comments in code) and tests for cache.

elbotho commented:
6 linked issues and the amount of discussion is probably a good indicator that this pr is to big :)

kulla reviewed (approved):
@elbotho Can you review my changes? I only want to merge master in it and update the set-uuid-state endpoint. The leftover comments are for follow ups.

elbotho commented on src/model/serlo.ts:
```diff
@@ -454,14 +457,84 @@ export function createSerloModel({
+        })
+
+      return value
     },
```
just to make sure, should we invalidate the cache of the contents uuid as well here?

kulla replied:
you mean the uuid the threads belongs to? It does not change.

elbotho commented on src/model/serlo.ts:
```diff
@@ -454,14 +457,84 @@ export function createSerloModel({
+      if (value !== null) {
+        await environment.cache.set<ThreadsPayload>({
+          key: getThreadIds._querySpec.getKey({ id: payload.objectId }),
+          getValue(current) {
```
that's pretty nice :)

elbotho commented on src/model/serlo.ts:
```diff
@@ -130,23 +135,23 @@ export function createSerloModel({
             await environment.cache.set({
-              key: `de.serlo.org/api/uuid/${uuidId}`,
+              key: getUuid._querySpec.getKey({ id }),
               value,
```
Just to make sure: Is the endpoint returning all notifications for the user when updating one?

elbotho commented:
@kulla I reviewed the parts that are not tests for now (see comments).
Imho: Go ahead and merge, reviewing the changes in the tests would take me more time and I'm not sure how useful my review would be there anyway :) I'll have a look out of interest.

kulla commented on src/model/serlo.ts:
```diff
@@ -454,14 +457,84 @@ export function createSerloModel({
+            if (current === undefined) return Promise.resolve(undefined)
+
+            current.firstCommentIds.push(value.id)
+            current.firstCommentIds.sort()
```
This is for the case more than one mutation is triggered at the same moment but the first mutation is delayed a little bit. I have moved the sort algorithm to the resolvers.
